llmfoundry/models/layers/attention.py

```diff
@@ -59,6 +66,7 @@ def scaled_multihead_dot_product_attention(
     attn_weight = q.matmul(k) * softmax_scale
 
     if attn_bias is not None:
+        attn_bias = attn_bias[:, :, -s_q:, -s_k:]
```

Comment:
@sashaDoubov I have some vague memory that torch compile blew up at negative indexing like this. Do you remember the details?

Reply:
Yes I got a scary error of the form:
```[41](https://mosaicml.slack.com/archives/D056APM0WB0/p1683927713531659)
    return (as_strided(buf0, (4, 2048, 2048), (12582912, 6144, 1)), as_strided(buf0, (4, 2048, 2048), (12582912, 6144, 1), 2048), as_strided(buf0, (4, 2048, 2048), (12582912, 6144, 1), 4096), as_strided(primals_2, (1, 16, 2048, 2048), (32768, 2048, 2048, 1), -4192256), buf0, as_strided(primals_1, (8192, 2048), (2048, 1)), as_strided(primals_3, (6144, 2048), (2048, 1)), )
RuntimeError: Tensor: invalid storage offset -4192256
```
This seemed to arise when dims 2 and 3 were smaller than s_q and s_k, and was trying to access elements in the attn_bias tensor with large negative indices.
The fix was clamping, something like:
```
  # necessary change for torch 2.0, otherwise there's a RuntimeError with negatives
  clamped_query_dim = max(0, attn_bias.size(2) - query.size(1))
  clamped_key_dim = max(0, attn_bias.size(3) - key.size(1))

  attn_bias = attn_bias[:, :, clamped_query_dim:, clamped_key_dim:]
```

Reply:
i'm not sure this is an issue
(i think u had a bug)

if `-s_q` or `-s_k` was `-4192256`, when `max(query.size(1)) = max_seq_len` and `max(key.size(1)) = max_seq_len`...
it seems like there was a bug somewhere...

or am I trippin?

@sashaDoubov is there a min repro of how i could hit this error?

Reply:
python handles the negative indexing, but compile doesn't like it.

Reply:
i dont recall why we were in a situation where one of those was highly negative, but i think it wasn't a bug?

Reply:
Have discussed this with both @dakinggg and @vchiley, don't think this is an issue for our code, I was able to narrow down the issue to some manual manipulation that I was performing in a different `if` branch of the code. This was reproduced in this file here, but _shouldn't_ be an issue here. https://github.com/sashaDoubov/llm-foundry/blob/repro_tensor_issue/llmfoundry/models/layers/min_repro_2_failing.py

Reply:
doesn't have to be part of this PR, but at some point we'll add torch.compile tests

Reply:
ok sounds like it should be fixed